File: cocos/assets/Script/comp/Bullet.ts
/**
 * Copyright 2023. Huawei Technologies Co., Ltd. All rights reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import {BulletData} from "../function/BulletList";
import {CmdType, CollideTag, Direction, GameSceneType} from "../function/FrameSync";
import global from "../../global";

const {ccclass} = cc._decorator;
/**
 * 子弹的逻辑
 */
@ccclass
export default class Bullet extends cc.Component {

    bulletId: number = 0;

    playerId: string = "";

    // 子弹飞行任务
    private flyTask = null;
    // 子弹飞行时的bullet信息
    private bulletData: BulletData;

    start() {
        global.room.onStopFrameSync(() => this.onStopFrameSync());
    }

    public onStopFrameSync() {
        clearInterval(this.flyTask);
    }

    public initBullet(bullet: BulletData) {
        this.node.name = bullet.bulletId.toString();
        this.bulletId = bullet.bulletId;
        this.playerId = bullet.playerId;
        this.node.x = bullet.x;
        this.node.y = bullet.y;
        this.bulletData = bullet;
        if(global.gameSceneType == GameSceneType.FOR_GAME && !global.isRequestFrameStatus && bullet.playerId == global.playerId){
            this.flyTask = setInterval(() => {
                this.updateBulletPos(this.bulletData);
            }, 200);
        }
    }

    updatePos(bullet: BulletData) {
        this.node.x = bullet.x;
        this.node.y = bullet.y;
    }

    destroyBullet(){
        clearInterval(this.flyTask);
        this.node.destroy();
    }

    /**
     * 碰撞检测
     * @param other
     */
    onCollisionEnter (other, self) {
        // 子弹碰到发射自己的飞机，不销毁
        if(other.tag == CollideTag.plane && other.node.name != this.playerId){
            console.log(`Bullet onCollisionEnter bulletId: ${this.bulletId}, selfTag: ${self.tag}, otherTag: ${other.tag}`);
            clearInterval(this.flyTask);
            this.node.destroy();
        }
    }


    // 更新子弹位置
    updateBulletPos(bullet: BulletData) {
        if(!bullet){
            return;
        }
        // 计算移动后的 x、y
        let x: number = bullet.x;
        let y: number = bullet.y;
        switch (bullet.direction) {
            case Direction.up:
                y = Math.min(global.bulletMaxY, bullet.y + global.bulletStepPixel);
                break;
            case Direction.down:
                y = Math.max(global.bulletSize, bullet.y - global.bulletStepPixel);
                break;
            case Direction.left:
                x = Math.max(global.bulletSize, bullet.x - global.bulletStepPixel);
                break;
            case Direction.right:
                x = Math.min(global.bulletMaxX, bullet.x + global.bulletStepPixel);
                break;
            // no default
        }
        // 子弹销毁
        if((bullet.direction == Direction.up && y == global.bulletMaxY) ||
            (bullet.direction == Direction.down && y == global.bulletSize) ||
            (bullet.direction == Direction.left && x == global.bulletSize) ||
            (bullet.direction == Direction.right && x == global.bulletMaxX)) {
            let frameData: string = JSON.stringify({
                cmd: CmdType.bulletDestroy,
                playerId: bullet.playerId,
                bulletId: bullet.bulletId,
            });
            console.log('----sendBulletDestroyFrame---' + frameData);
            this.bulletData.x = x;
            this.bulletData.y = y;
            this.bulletData.needDestroy = true;
            global.room.sendFrame(frameData);
            clearInterval(this.flyTask);
        }
        // 子弹继续飞行
        else {
            let frameData: string = JSON.stringify({
                cmd: CmdType.bulletFly,
                playerId: bullet.playerId,
                bulletId: bullet.bulletId,
                x,
                y,
                direction:bullet.direction
            });
            console.log('----sendBulletFlyFrame---' + frameData);

            this.bulletData.x = x;
            this.bulletData.y = y;
            global.room.sendFrame(frameData);
        }
    }
}

File: unity/Assets/Script/Comp/Bullet.cs
/**
 * Copyright 2023. Huawei Technologies Co., Ltd. All rights reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

using System.Linq;
using Com.Huawei.Game.Gobes.Utils;
using UnityEngine;
public class Bullet : MonoBehaviour
{
    public int speed = 20;
    public bool isDie = true;
    public int bullectId = 0;
    public string playerId = "";
    
    public void initBullet(int x,int y,string PlayerId,int BullectId) {
        playerId = PlayerId;
        bullectId = BullectId;
        // 如果宽高超过边界，销毁子弹
        if(x > FrameSync._maxX || x < FrameSync._minX || y > FrameSync._maxY || y < FrameSync._minY){
            // 子弹数据销毁
            FrameSync.frameSyncBulletList = FrameSync.frameSyncBulletList.Where(item => !(item.playerId.Equals(PlayerId) && item.bulletId == BullectId)).ToList();
        }else{
            gameObject.transform.position = new Vector3(x, y, 0);
        }

    }
    
    /**
     * 碰撞检测
     * @param collision
     */
    public void OnTriggerEnter(Collider other)
    {
        Debug.Log("子弹碰撞2-----------------------------"+other.tag+"==="+other.gameObject.tag);
        if (FrameSync.PlayerTag.Equals(other.tag))
        {
            CollisionFrameList<FrameSync.CollisionFrameData>.CollisionFrameData<FrameSync.CollisionFrameData> collisionFrameData =
                new CollisionFrameList<FrameSync.CollisionFrameData>.CollisionFrameData<FrameSync.CollisionFrameData>();
            collisionFrameData.state = new FrameSync.CollisionFrameData();
            collisionFrameData.state.cmd = FrameSync.FrameSyncCmd.collide;
            collisionFrameData.otherTag = other.tag;//被碰撞体标签
            collisionFrameData.selfTag = FrameSync.BulletTag;//碰撞体标签
            collisionFrameData.playerId = playerId;
            collisionFrameData.bulletId = bullectId;
            
            string frameData = CommonUtils.JsonSerializer(collisionFrameData);
            // 调用SDK发送帧数据
            string[] frameDatas = new string[] { frameData };
            Global.Room.SendFrame(frameDatas, response => {});
        }
    }
}
